evm/src/main/java/org/hyperledger/besu/evm/precompile/AltBN128AddPrecompiledContract.java

```diff
@@ -74,11 +78,24 @@ public long gasRequirement(final Bytes input) {
   @Override
   public PrecompileContractResult computePrecompile(
       final Bytes input, @Nonnull final MessageFrame messageFrame) {
+
+    PrecompileInputResultTuple res;
+
+    if (enableResultCaching) {
+      res = bnAddCache.getIfPresent(input.hashCode());
```

Comment:
As discussed before, hashCode is not a good key here as you may have collisions. We need to use either the whole input or a hashing function that assumes no collisions.

Reply:
I was expecting and embracing collisions.  I think a cache limited to <=1000 items is unlikely to have collisions in the int space unless it is an intentional attempt to create collisions.  In that case, the equals() check will prevent using the cache and it will be overwritten with the new computed value.

If we use input Bytes, it seems we can attack the cache by filling it with collisions and cause a worst case performance where we have to use byte-by-byte equals() method to distinguish between the hashCode collisions.  

For that reason I think leaning into a 1 cache item per hash code is a safer strategy.   It will work well with few naturally occurring collisions, and won't result in the ability to attack the cache and cause slow blocks.

Reply:
In another caching test PR I've used [xxHash](https://xxhash.com/) to create the cache key, which is supposed to be very fast.

Here's the PR where I've used it:
https://github.com/daniellehrner/besu/commit/c75f975c3753fba4bd8346345c04462a5948baf7#diff-006f4a4dd9869817685273bb0bc45923b06e6a926661e7ac84615564576e9060R146

Reply:
Now that I understand that you have a second check on the input, I think it is a pretty smart method, I'm interested to see what the profiling show, basically what is the overhead of the equals on the input when there is a collision.
When there no collision, it is pretty good algorithm complexity. We can also play with the size of the cache to reduce collisions, and check if the memory overhead is not too big. This will depend on the input size, which not fixed I guess.